<a href="https://colab.research.google.com/github/vs1991/NLP-Basics/blob/main/CustomerComplaint.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
from google.colab import drive
drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [3]:
import os
os.environ['KAGGLE_CONFIG_DIR'] = "/content/gdrive/My Drive/Data Science -NLP"


In [4]:
#changing the working directory
%cd /content/gdrive/My Drive/Data Science -NLP
#Check the present working directory using pwd command

/content/gdrive/My Drive/Data Science -NLP


In [5]:
ls

'Author Iden Test.csv'  'Author Iden Train.csv'   Consumer_Complaints.csv


In [6]:
import pandas as pd 
import numpy as np
import warnings
warnings.filterwarnings("ignore")

In [7]:
full_data=pd.read_csv('Consumer_Complaints.csv')
full_data.head(3)

Date received           Product  ... Consumer disputed? Complaint ID
0     3/12/2014          Mortgage  ...                 No       759217
1     10/1/2016  Credit reporting  ...                 No      2141773
2    10/17/2016     Consumer Loan  ...                 No      2163100

[3 rows x 18 columns]

In [7]:
print(full_data.shape)

(903983, 18)


Since our data contains 903983 complaints,I will consider only consider only few complaints in our data analysis

**Our goal is to send the complaints to the right vertical**

In [22]:
data=full_data[['Consumer complaint narrative','Product']][0:20000]
data.dropna(inplace=True)
data.columns=['X','y']
data.head(3)

X                 y
1  I have outdated information on my credit repor...  Credit reporting
2  I purchased a new car on XXXX XXXX. The car de...     Consumer Loan
7  An account on my credit report has a mistaken ...  Credit reporting

In [23]:
data.shape

(4218, 2)

In [24]:
print(data.iloc[2][0])

An account on my credit report has a mistaken date. I mailed in a debt validation letter to allow XXXX to correct the information. I received a letter in the mail, stating that Experian received my correspondence and found it to be " suspicious '' and that " I did n't write it ''. Experian 's letter is worded to imply that I am incapable of writing my own letter. I was deeply offended by this implication. 
I called Experian to figure out why my letter was so suspicious. I spoke to a representative who was incredibly unhelpful, She did not effectively answer any questions I asked of her, and she kept ignoring what I was saying regarding the offensive letter and my dispute process. I feel the representative did what she wanted to do, and I am not satisfied. It is STILL not clear to me why I received this letter. I typed this letter, I signed this letter, and I paid to mail this letter, yet Experian willfully disregarded my lawful request. 
I am disgusted with this entire situation, and I

**Word combine to from sentence ,sentence combine to form paragraphs.**
First,we need to convert the text data(customer complaint narrative)into numerical data,since ML model can't read text data.**How do we do that ?**

Here comes NLP models.Before that let's go through some of the difficulty of using text data.



1.   **Synonymy:-**   different words having the same meaning
2.   **Ambiguity :-** The bank deposit rate is quite high" and "He stood near the bank admiring the river". In these statements, the word bank has completely different meanings.
3.   **Anaphora Resolution:-**George is my friend. He likes football". In the second statement he refers to George. It is difficult for the computers to discern what person/entity the pronoun he is referring to.
4.   **Language related issues :-**Every language has its own uniqueness
5.   **Out of Vocabulary problem:-**Machines have a hard time adapting to any new constructs that humans come up with. As humans when we come across a word we haven't seen earlier, we might not understand its meaning instantly. But this does not mean we cannot adapt. After looking at the word in several different sentences and understanding its usage, we understand the context and meaning of the new word. Machines can only handle data that they have seen before. It is unable to adapt well.
6.   **Language generation:-**While language understanding is hard, language generation too has its own set of challenges. For chatbots to work effectively, they need to communication properly constructed sentences which are grammatically correct. This is quite a hard problem and a challenge that needs to be overcome. 







**Use cases of NLP**

1. **Sentiment Analysis -** Finding if the text is leaning towards a positive or negative sentiment.
2. **Text Classification -** Categorizing text to various categories
3. **Document Summarization -** Compressing a paragraph/document into few words or sentences
4. **Parts of Speech Tagging-** Figuring out the various nouns, adverbs, verbs etc in the text
5.  **Machine translation -** Translate text from one language to another
6. **Named Entity Recognition -** Identify the entities present in text
7. **Conversational AI -** Chat with a machine in natural language and get queries resolved


To answer our question how to convert text data into numerical data 
***For that, we need to arrive at a fundamental component of text known as tokens.***


Tokenization is the process of splitting the text into smaller parts called tokens. Tokens are the basic units of a particular dataset. The choice of tokens could be based on the application we are working on.

In [25]:
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [26]:
#text pre-processing library
#first step is to breakdown sentences into words ,we will use the nltk library for that 
import nltk 
from nltk.tokenize import word_tokenize 
# we will drop na values from the data 
#data.dropna(inplace=True)
print('New Shape of data is ',data.shape,'\n')
print('let us look at our first complaint:-')
first_complaint=data.iloc[5][0]
print(first_complaint)

#let us tokenzie the data using split function of pandas 
bag_of_words_split=first_complaint.split()
print(bag_of_words_split,'\n')

#let us tokenzie the data using nltk word tokenize 
bag_of_words_wordt=word_tokenize(first_complaint)
print(bag_of_words_wordt,'\n')
print(first_complaint[23])

New Shape of data is  (4218, 2) 

let us look at our first complaint:-
Started the refinance of home mortgage process with cash out option on XX/XX/2016. Necessary documents were submitted by XXXX. After initial review, got good faith estimate with loan amount and closing cost. Based on this estimate, a deposit of {$350.00} was made towards appraisal. Appraisal came with lesser amount by {$5000.00}. Agreed to reduce the loan amount to that extent. However, got a revised estimate which was less by {$30000.00} and with additional closing cost towards points etc. In between got numerous revised estimates with different loan amounts and closing cost. It took more than 2 months to reach any definite closing document. Hence, want to get back the deposit of {$350.00}.
['Started', 'the', 'refinance', 'of', 'home', 'mortgage', 'process', 'with', 'cash', 'out', 'option', 'on', 'XX/XX/2016.', 'Necessary', 'documents', 'were', 'submitted', 'by', 'XXXX.', 'After', 'initial', 'review,', 'got', 'good

One could also tokenize a paragraph into constituent sentences. To split into definite sentences

The importance of converting words to lower case - All words should be converted to lowercase while doing NLP. The reason behind being, that "Yorkshire" and "yorkshire" even though are the same word, will be considered 2 separate words while converting the words into numbers.

In [27]:
from nltk.tokenize import sent_tokenize

#sentence tokenizing using sent_tokenize
list_of_sentences=sent_tokenize(first_complaint)
print(list_of_sentences,'\n')

#lowering first complaint 
first_complaint_lower=first_complaint.lower()
print(first_complaint_lower,'\n')

#worde tokenizing the lower case complaints 
bag_of_words_lower=word_tokenize(first_complaint_lower)
print(bag_of_words_lower)

['Started the refinance of home mortgage process with cash out option on XX/XX/2016.', 'Necessary documents were submitted by XXXX.', 'After initial review, got good faith estimate with loan amount and closing cost.', 'Based on this estimate, a deposit of {$350.00} was made towards appraisal.', 'Appraisal came with lesser amount by {$5000.00}.', 'Agreed to reduce the loan amount to that extent.', 'However, got a revised estimate which was less by {$30000.00} and with additional closing cost towards points etc.', 'In between got numerous revised estimates with different loan amounts and closing cost.', 'It took more than 2 months to reach any definite closing document.', 'Hence, want to get back the deposit of {$350.00}.'] 

started the refinance of home mortgage process with cash out option on xx/xx/2016. necessary documents were submitted by xxxx. after initial review, got good faith estimate with loan amount and closing cost. based on this estimate, a deposit of {$350.00} was made to

# Stemming and Lemmatization

**Stemming** Stemming is the process of converting the words of a sentence to its non-changing portions. So stemming a word or sentence may result in words that are not actual words. Stems are created by removing the suffixes or prefixes used with a word.

For eg: Likes, liked, likely, unlike⇒like

 The popular ones include:

1. Porter Stemmer(Implemented in almost all languages)
2. Paice Stemmer
3. Lovins Stemmer


**Lemmatization:**

This method is a more refined way of breaking words through the use of a vocabulary and morphological analysis of words. The aim is to always return the base form of a word known as lemma.

Consider the following words:

'Studied', 'Studious' ,'Studying'

Stemming of them will result in Studi

Lemmatisation of them will result in Stu

**Bag of words:**

The problem with modeling text is that there is no well defined fixed-length inputs.

A bag of words model is a way of extracting features from text for use in modeling. In this approach, we use the tokenized words for each observation and find out the frequency of each token.

This process of converting text data to numbers is called vectorization

There are multiple methods to convert words to numbers. We will be start with discussing the count Vectorizer.



In [28]:
from collections import Counter
count_vectorizer=Counter(bag_of_words_lower)
print(count_vectorizer)

Counter({'.': 10, 'with': 5, ',': 4, 'closing': 4, '{': 4, '$': 4, '}': 4, 'to': 4, 'the': 3, 'of': 3, 'by': 3, 'got': 3, 'estimate': 3, 'loan': 3, 'amount': 3, 'and': 3, 'cost': 3, 'on': 2, 'a': 2, 'deposit': 2, '350.00': 2, 'was': 2, 'towards': 2, 'appraisal': 2, 'revised': 2, 'started': 1, 'refinance': 1, 'home': 1, 'mortgage': 1, 'process': 1, 'cash': 1, 'out': 1, 'option': 1, 'xx/xx/2016': 1, 'necessary': 1, 'documents': 1, 'were': 1, 'submitted': 1, 'xxxx': 1, 'after': 1, 'initial': 1, 'review': 1, 'good': 1, 'faith': 1, 'based': 1, 'this': 1, 'made': 1, 'came': 1, 'lesser': 1, '5000.00': 1, 'agreed': 1, 'reduce': 1, 'that': 1, 'extent': 1, 'however': 1, 'which': 1, 'less': 1, '30000.00': 1, 'additional': 1, 'points': 1, 'etc': 1, 'in': 1, 'between': 1, 'numerous': 1, 'estimates': 1, 'different': 1, 'amounts': 1, 'it': 1, 'took': 1, 'more': 1, 'than': 1, '2': 1, 'months': 1, 'reach': 1, 'any': 1, 'definite': 1, 'document': 1, 'hence': 1, 'want': 1, 'get': 1, 'back': 1})


In [39]:
from sklearn.metrics import accuracy_score,roc_auc_score
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.preprocessing import LabelEncoder

#Code starts here

#Subsetting 'X'
all_text = data[["X"]]

#Converting 'X' to lower case
all_text["X"] = all_text['X'].str.lower()

#Initialising a count vectorizer object
cv = CountVectorizer()

#Creating the count vectorizer of our 'X' column
vector =cv.fit_transform(all_text["X"])

#Converting the count vectoriser to array
X = vector.toarray()

#Subsetting y
labels = data[["y"]]

#Initialising a label encoder object
le = LabelEncoder()

#Label encoding 'y' column
labels["y"] = le.fit_transform(labels["y"])

#Splitting the dataset into train and test
X_train,X_test,y_train,y_test = train_test_split(X,labels["y"],test_size=0.4,random_state=42)

#Initialising Logistic Regression model
log_reg = LogisticRegression(random_state=42)

#Fitting the model on train data
log_reg.fit(X_train,y_train)

#Finding the accuracy score on test data
acc = log_reg.score(X_test,y_test)
print ('Accuracy with stopwords',acc)

#Code ends here

Accuracy with stopwords 0.7292654028436019


**Removing Stopwords**


we have seen 49% accuracy of predicting the product category. Now the question we need to ask is - can we improve the accuracy further? The answer lies in dealing with stopwords.


**Stop Words**

count vectorizer is essentially a Ranking algorithm, in the way that it gives a higher weight to words which have appeared more number of times. 
 It does not assign any importance to the order in which the words have appeared in the sentences.
Another point to be considered is the fact that words like "a","an","the" etc. will appear more number of times than the rest of the words as they are common articles. Using a Count vectorizer out of the box on a paragraph or a body of text will invariably give the highest count to these common words.

**One way to rectify this, is to remove these commonly occurring words.**

In [40]:
cv_stop=CountVectorizer(stop_words='english')

#creating a count vectorizer of our x column
vector_stop=cv_stop.fit_transform(all_text['X'])

#converting the count vectorizer to array 
X_stop=vector_stop.toarray()

#Splitting the data to train and test
X_train,X_test,y_train,y_test = train_test_split(X_stop,labels["y"],test_size=0.4,random_state=42)

#Initalising a logistic regression model
log_reg = LogisticRegression(random_state=42)

#Fitting the model on train
log_reg.fit(X_train,y_train)

#Finding the accuracy score on test data
stop_acc = log_reg.score(X_test,y_test)
print ('Accuracy without stopwords',stop_acc)

Accuracy without stopwords 0.7470379146919431



a count vectorizer, counts the occurences of the words in a document and all the documents are considered independent of each other.However in cases where multiple documents are involved, count vectorizer still does not assume any interdependence between the documents and considers each of the documents as a seperate entity.


**It does not rank the words based on their importance in the document, but just based on whether they exist or not.**

In fact, the process of converting, text to numbers should essentially be a ranking system of the words so that the documents can each get a score based on what words they contain. All words cannot have the same imprtance or relevance in the document right?
There are two ways to approach document similarity:

1. TF-IDF Score

2. Cosine Similarity




# 1. Tfidf(Term Frequency and Inverse Document Frequency)  with examples
is the holy grail of ranking metrics to convert text to numbers 



**The ranking system in a count vectorizer is purely occurence based on a single document only!**

TF-IDF takes it a step further and ranks the words based not just on their occurences in one document but across all the documents.
if CV or Count vectorizer was giving more importance to words because they have appeared multiple times in the document, TF-IDF will rank them high if they have appeared only in that document, meaning that they are rare, hence higher importance and lower if they have appeared in all or most documents, because they are more common, hence lower ranking.

Consider a scenario where there are 5 documents and all are talking about football. The word football would have appeared multiple times in each document. CV is going to rank football consistently high and infact give the word football a different value across all 5 documents based on how many times that word has appeared in that document. In other words, it is assuming, that the more number of times a word appears, the more important it is. That is exactly what the TF or the Term Frequency component in TF-IDF does.

**IDF** on the other hand now is the dominating factor in TFIDF which is going to find out the number of times football has also appeared in the other 4 documents except for the one it is currently seeing. If football has also appeared in rest of the documents, it means that though football is important to that one document based on the number of occurences, considering it has appeared in the rest as well, it is not that rare or more common, hence the importance now is going to reduce instead of going high! italicised text


**The ranking system is across the entire corpus or all documents. It is not a single document based metric!**





**How is Tfidf calculate?**


the tfidf constitues of 2 parts :

the 1st computes the normalized **Term Frequency(Tf)** a.k.a the no. of times the word has appeared in a document,divided by the total no. of words in the document.

the 2nd term **Inverse Document Frequency(IDF)** computes the logarithm of the no. of the documents to the number of times the word appears in the different documents.


**TF:** Term Frequency, which measures how frequently a term occurs in a document. Since every document is different in length, it is possible that a term would appear much more times in long documents than shorter ones. Thus, the term frequency is often divided by the document length (aka. the total number of terms in the document) as a way of normalization:

TF(t) = (Number of times term t appears in a document) / (Total number of terms in the document)


**IDF:** Inverse Document Frequency, which measures how important a term is. While computing TF, all terms are considered equally important. However it is known that certain terms, such as "is", "of", and "that", may appear a lot of times but have little importance. Thus we need to weigh down the frequent terms while scale up the rare ones, by computing the following:

IDF(t) = log_e(Total number of documents / Number of documents with term t in it).

In [42]:
complaint_1=data['X'].iloc[0]
complaint_2=data['X'].iloc[1]
complaint_3=data['X'].iloc[2]

print ("Complaint 1: ", complaint_1)

print ("\nComplaint 2: ", complaint_2)

print ("\nComplaint 3: ", complaint_3)

Complaint 1:  I have outdated information on my credit report that I have previously disputed that has yet to be removed this information is more then seven years old and does not meet credit reporting requirements

Complaint 2:  I purchased a new car on XXXX XXXX. The car dealer called Citizens Bank to get a 10 day payoff on my loan, good till XXXX XXXX. The dealer sent the check the next day. When I balanced my checkbook on XXXX XXXX. I noticed that Citizens bank had taken the automatic payment out of my checking account at XXXX XXXX XXXX Bank. I called Citizens and they stated that they did not close the loan until XXXX XXXX. ( stating that they did not receive the check until XXXX. XXXX. ). I told them that I did not believe that the check took that long to arrive. XXXX told me a check was issued to me for the amount overpaid, they deducted additional interest. Today ( XXXX XXXX, ) I called Citizens Bank again and talked to a supervisor named XXXX, because on XXXX XXXX. I received 

In [50]:
from sklearn.feature_extraction.text import TfidfVectorizer
# list of text documents called sents
sents = [complaint_1, complaint_2, complaint_3]
# create the transform
vectorizer = TfidfVectorizer()

vectorizer.fit(sents)

vector = vectorizer.transform(sents)

print("Shape of the vectorized sentence:",vector.shape)

vector_values = vector.toarray().tolist()[0]

print("The tf-idf score of first five elements:",vector_values[:20])


# Converting the tf-idf score with the word into a dictionary
import operator
sorted_x=sorted(vectorizer.vocabulary_.items(),key=operator.itemgetter(1))

words=[x[0] for x in sorted_x]
d=dict(zip(words,vector_values))
print("Dictionary of words with tf-idf score:\n", d)
print("Sorted dictonary:\n")
print (sorted(d.items(), key=operator.itemgetter(1), reverse = True))

Shape of the vectorized sentence: (3, 223)
The tf-idf score of first five elements: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.11807902273813668, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
Dictionary of words with tf-idf score:
 {'10': 0.0, 'about': 0.0, 'account': 0.0, 'additional': 0.0, 'afraid': 0.0, 'after': 0.0, 'again': 0.0, 'allow': 0.0, 'am': 0.0, 'amount': 0.0, 'an': 0.0, 'and': 0.11807902273813668, 'another': 0.0, 'answer': 0.0, 'any': 0.0, 'appropriately': 0.0, 'are': 0.0, 'arrive': 0.0, 'asked': 0.0, 'at': 0.0, 'automatic': 0.0, 'balanced': 0.0, 'bank': 0.0, 'be': 0.11807902273813668, 'because': 0.0, 'been': 0.0, 'believe': 0.0, 'business': 0.0, 'but': 0.0, 'by': 0.0, 'called': 0.0, 'car': 0.0, 'check': 0.0, 'checkbook': 0.0, 'checking': 0.0, 'citizens': 0.0, 'clear': 0.0, 'close': 0.0, 'consumer': 0.0, 'contact': 0.0, 'correct': 0.0, 'correspondence': 0.0, 'could': 0.0, 'credit': 0.3040964814250496, 'date': 0.0, 'dated': 0.0, 'day': 0.0, 'days': 0.0, 'dealer': 

In [51]:
sents=[]
for x in range(100):
    sents.append(data["X"].iloc[x])

from sklearn.feature_extraction.text import TfidfVectorizer
# list of text documents called sents
# create the transform
vectorizer = TfidfVectorizer()
# tokenize and build vocab
vectorizer.fit(sents)
vector = vectorizer.transform(sents)
vector.shape 
vector_values = vector.toarray().tolist()[0]

sorted_x = sorted(vectorizer.vocabulary_.items(), key=operator.itemgetter(1))
words = [x[0] for x in sorted_x]
d = dict(zip(words,vector_values))
print("Sorted dictionary: \n")
print ((sorted(d.items(), key=operator.itemgetter(1), reverse = True))[:20])

Sorted dictionary: 

[('outdated', 0.30698816696178366), ('requirements', 0.30698816696178366), ('seven', 0.28169892181541933), ('meet', 0.26375591683431887), ('information', 0.25693994145225296), ('previously', 0.2498382409130902), ('disputed', 0.23846667168795455), ('old', 0.22052366670685408), ('yet', 0.21317742654159022), ('credit', 0.20755516271607583), ('does', 0.19523442156048976), ('more', 0.19523442156048976), ('removed', 0.19024208161296743), ('reporting', 0.1856198967748199), ('have', 0.1669613523183918), ('then', 0.15742915392599566), ('that', 0.15697667242334717), ('years', 0.15465665513790783), ('report', 0.1340591664519245), ('has', 0.1302779510221311)]


# Tfidf On our dataframe

In [57]:
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf=TfidfVectorizer(stop_words='english')

vector_tfidf=tfidf.fit_transform(data['X'])

#converting this to array 
X_tfidf=vector_tfidf.toarray()

X_train,X_test,y_train,y_test = train_test_split(X_tfidf,labels["y"],test_size=0.4,random_state=42)

#Initalising a logistic regression model
log_reg = LogisticRegression(random_state=42)

#Fitting the model on train
log_reg.fit(X_train,y_train)

#Finding the accuracy score on test data
tfidf_acc = log_reg.score(X_test,y_test)
print ('Accuracy with Tfidf Vectorizer',tfidf_acc)

Accuracy with Tfidf Vectorizer 0.7571090047393365


# Naive Bayes Classifier

In [59]:
from sklearn.naive_bayes import MultinomialNB

all_text = data[["X"]]

# Converting the 'X' column to lower case
all_text["X"] = all_text['X'].str.lower()

# Initialising a tfidf vectorizer object with stopwords
tfidf = TfidfVectorizer(stop_words="english")

# Vectorizing the 'X' column
vector = tfidf.fit_transform(all_text["X"])

# Converting vector to array
X_tfidf = vector.toarray()

# y

# Subsetting 'y' column
labels = data[["y"]]

# Initialising label encoder object
le = LabelEncoder()

# Label encoding 'y' column
labels["y"] = le.fit_transform(labels["y"])

# Splitting the data into train and test
X_train, X_test, y_train, y_test = train_test_split(
    X_tfidf, labels["y"], test_size=0.4, random_state=42)

# Initialsing a naive bayes classifier
nb = MultinomialNB()

# Fitting the model on train data
nb.fit(X_train, y_train)

# Finding the accuracy score of model on test data
nb_acc = nb.score(X_test, y_test)
print(nb_acc)

0.6143364928909952


In [60]:
from sklearn.metrics import classification_report

print (classification_report(y_test,nb.predict(X_test)))

              precision    recall  f1-score   support

           0       1.00      0.16      0.28       154
           1       0.00      0.00      0.00       107
           2       0.72      0.32      0.44       193
           3       0.76      0.80      0.78       310
           4       0.50      0.94      0.66       397
           5       0.00      0.00      0.00        22
           6       0.64      0.91      0.75       344
           7       0.00      0.00      0.00         3
           8       0.00      0.00      0.00        21
           9       0.00      0.00      0.00        16
          10       0.93      0.12      0.21       121

    accuracy                           0.61      1688
   macro avg       0.41      0.30      0.28      1688
weighted avg       0.63      0.61      0.54      1688



In [63]:
from imblearn.over_sampling import RandomOverSampler
from sklearn.naive_bayes import MultinomialNB

ros=RandomOverSampler(random_state=42)
X_ros,y_ros=ros.fit_sample(X_train,y_train)
nb = MultinomialNB()
# Fitting the model on train data
nb.fit(X_ros, y_ros)

# Finding the accuracy score of model on test data
nb_acc = nb.score(X_test, y_test)
print(nb_acc)

0.773696682464455


In [64]:
from sklearn.metrics import classification_report

print (classification_report(y_test,nb.predict(X_test)))

              precision    recall  f1-score   support

           0       0.68      0.80      0.73       154
           1       0.71      0.56      0.62       107
           2       0.71      0.72      0.72       193
           3       0.72      0.83      0.77       310
           4       0.85      0.69      0.76       397
           5       0.50      0.59      0.54        22
           6       0.90      0.90      0.90       344
           7       0.00      0.00      0.00         3
           8       0.52      0.57      0.55        21
           9       0.48      0.62      0.54        16
          10       0.80      0.88      0.84       121
          11       0.00      0.00      0.00         0

    accuracy                           0.77      1688
   macro avg       0.57      0.60      0.58      1688
weighted avg       0.78      0.77      0.77      1688



In [68]:
from sklearn.svm import SVC

svc=SVC(random_state=42)
svc.fit(X_ros,y_ros)


SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='rbf',
    max_iter=-1, probability=False, random_state=42, shrinking=True, tol=0.001,
    verbose=False)

In [69]:
# Finding the accuracy score of model on test data
nb_acc = svc.score(X_test, y_test)
print(nb_acc)

0.7565165876777251
